Gems/ROS2/Code/Source/VehicleDynamics/DriveModels/SkidSteeringDriveModel.cpp

```diff
@@ -38,72 +38,95 @@ namespace ROS2::VehicleDynamics
     void SkidSteeringDriveModel::Activate(const VehicleConfiguration& vehicleConfig)
     {
         m_config = vehicleConfig;
-        m_wheelsData.clear();
-        m_wheelColumns.clear();
+        m_wheelsCache.clear();
+        m_initialized = false;
     }
 
-    AZStd::tuple<VehicleDynamics::WheelControllerComponent*, AZ::Vector2, AZ::Vector3> SkidSteeringDriveModel::ProduceWheelColumn(
-        int wheelNumber, const AxleConfiguration& axle, const int axisCount) const
+    void SkidSteeringDriveModel::InitializeCache()
     {
-        const auto wheelCount = axle.m_axleWheels.size();
-        const auto wheelEntityId = axle.m_axleWheels[wheelNumber];
+        // Compute number of drive axles to scale the contribution of each wheel to vehicle's velocity
+        int driveAxlesCount = 0;
+        for (const auto& axle : m_config.m_axles)
+        {
+            if (axle.m_isDrive && axle.m_axleWheels.size() > 1)
+            {
+                driveAxlesCount++;
+            }
+            AZ_Warning(
+                "SkidSteeringDriveModel",
+                axle.m_axleWheels.size() > 1,
+                "Axle %s has not enough wheels (%d)",
+                axle.m_axleTag.c_str(),
+                axle.m_axleWheels.size());
+        }
+        AZ_Warning("SkidSteeringDriveModel", driveAxlesCount != 0, "Skid steering model does not have any drive wheels.");
+
+        // Find the contribution of each wheel to vehicle's velocity
+        for (const auto& axle : m_config.m_axles)
+        {
+            const auto wheelCount = axle.m_axleWheels.size();
+            if (!axle.m_isDrive || wheelCount <= 1)
+            {
+                continue;
+            }
+            for (size_t wheelId = 0; wheelId < wheelCount; wheelId++)
+            {
+                const auto column = ProduceWheelColumn(wheelId, axle, driveAxlesCount);
+                if (column.wheelControllerComponentPtr != nullptr)
+                {
+                    m_wheelsCache.emplace_back(AZStd::move(column));
+                }
+            }
+        }
+
+        m_initialized = true;
+    }
+
+    SkidSteeringDriveModel::SkidSteeringWheelData SkidSteeringDriveModel::ProduceWheelColumn(
```

Comment:
Consider move semantic for `SkidSteeringDriveModel` to optimize return. But since this is not executed periodically it's not crucial. Nice to have.

Reply:
Please correct me if I am wrong, but the temporary in `SkidSteeringDriveModel::ProduceWheelColumn` is implicitly treated as an rvalue in the return statement by the compiler's RVO.

Reply:
You are right! I have checked and according to the docs this is a perfect candidate for rvo.